### PR TITLE
Add support for Azure Developer CLI and Dev Container

### PR DESCRIPTION
**Pull request by pamelafox**
This PR adds support for using the Azure Developer CLI to provision the resources and deploy the code to App Service. The architecture is very similar to what's used in https://github.com/Azure-Samples/azure-search-openai-demo/ except that this project does not currently need blob storage. 

Overview of files added:

- `.devcontainer/Dockerfile` : Python image with azd
- `.devcontainer/devcontainer.json` : Add Azure CLI and node features
- `README_azd.md` : I added this as a separate readme instead of modifying the current readme, as I'm not sure if you want to keep the current deployment instructions which uses a Docker image. AZD seems preferable since developers can modify the code, but I don't want to presume, since I don't know exactly how this integrates with the Studio.
- `azure.yaml` : Needed by azd, describes what to deploy to app service and includes hooks for uploading docs to index.
- `infra/` : Contains all the Bicep files that describe the resources: App Service, App Service Plan, Cognitive Search, Form Recognizer (for doc upload). 
- `scripts/prepdocs.ps1` : Powershell script that runs prepdocs.py (copied from azure-search-openai-demo)
- `scripts/prepdocs.sh` : Shell script that runs prepdocs.py (copied from azure-search-openai-demo)
- `scripts/prepdocs.py` : An adaptation of data_preparation.py that uses the Azure SDK instead of raw HTTP requests. Makes calls to data_utils.py for parsing and chunking. In theory, you could remove data_preparation.py and config.json entirely, if you're comfortable with prepdocs.py instead. I noticed data_preparation.py can configure multiple indices, whereas prepdocs.py only attempts to configure one, since the app only uses one.
- `./start.sh` : A shell script version of start.cmd so that I can use it on my Mac / Dev Container.

A few observations:

* This repo currently uses keys instead of managed identity. General best practice is keys, so if possible, we should move over to that.
* This repo doesn't seem to integrate with Blob storage in anyway, even though that is mentioned as a possible data source for this feature (in the video). Is that pending?
* This repo still contains infrastructure/deployment.json and various files that create a Docker image. Not sure if we want to keep that or only recommend azd.
* The chunking code doesn't seem to work as well in this repo as it does in azure-search-openai-demo, I see it splitting inside sentences/words sometimes.
* In the UI, the citations sometimes seem wrong after the first message. I assume that's an unrelated issue to my change, but I can dig into it if you haven't seen that yourself. 
* The app code is currently in the root of this project, which means that azd has to upload everything in the repo, even the data. It's preferred to move app code into an "app" or "src" folder so that only that can be deployed. That's a larger change, however, so I left it out of this PR.

I've tested this on my Mac M1 in a Dev Container, and I have tested it in Codespaces. I do not have a Windows machine so it'd be helpful for someone to test prepdocs.ps1.

### CODE REVIEW

**Comment by pamelafox** [comment]
@sarah-widder Here's my PR for the initial changes. I'll look into automating the AAD authentication next.

**Comment by pamelafox** [comment]
Merged AAD PR into this one

File: app.py [hunk 1]
```diff
@@ -44,7 +44,7 @@ def static_file(path):
 SHOULD_STREAM = True if AZURE_OPENAI_STREAM.lower() == "true" else False
 
 def is_chat_model():
-    if 'gpt-4' in AZURE_OPENAI_MODEL_NAME.lower():
+    if 'gpt-4' in AZURE_OPENAI_MODEL_NAME.lower() or 'gpt-35' in AZURE_OPENAI_MODEL_NAME.lower():
```
Comment (sarah-widder) [inline app.py:47]: Let's revert this change - we want to keep using completions for Turbo for better AIQ in the with-data case for now. There was a bug with Turbo models named "chat" which has been fixed.

**Review by sarah-widder — APPROVED** [review]
Great contribution to the repo, thanks!

**Comment by pamelafox** [comment]
@sarah-widder I've changed STOP_SEQUENCE to default to an empty string, so this now works for me.
